{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# EX4: try/except, files and sub programs "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## ex 4.1\n",
    "\n",
    "Prepare on you PC in the SAME directory of our python program\n",
    "2 or more files composed by multiple integer values, comma separated\n",
    "like that: \n",
    "\n",
    "120,230,100,400,500  called \n",
    "samples_001.csv, samples_002.csv and so on,\n",
    "where 001 002 are the patient code.\n",
    "\n",
    "Write a program that receives in input a string representing the patent code, check if file exists, and prints the average of values,\n",
    "or False if file cannot be open or contents is invalid."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Enter patients code: 001\n",
      "270.0\n",
      "270.0\n"
     ]
    }
   ],
   "source": [
    "FILE_PREFIX = \"samples_\"\n",
    "FILE_SUFFIX = \".csv\"\n",
    "\n",
    "\n",
    "def buildName(patientCode):\n",
    "\tfname = FILE_PREFIX + patientCode + FILE_SUFFIX\n",
    "\treturn fname\n",
    "\n",
    "\n",
    "def avgFrom(line):\n",
    "\tvalues = line.split(',')\n",
    "\tvaluesCount = len(values)\n",
    "\tif valuesCount > 0:\n",
    "\t\tsum = int(values[0])\n",
    "\t\tfor i in range(1, valuesCount):\n",
    "\t\t\tvalue = int(values[i])\n",
    "\t\t\tsum += value\n",
    "\t\treturn float(sum) / valuesCount  # pass to float with right conversion\n",
    "\telse:\n",
    "\t\treturn 0.0\n",
    "\n",
    "def process(fileName):\n",
    "\ttry:\n",
    "\t\tf = open(fileName)\n",
    "\texcept:\n",
    "\t\treturn False\n",
    "\telse:\n",
    "\t\tlines = f.read().splitlines()\n",
    "\t\tf.close()\n",
    "\t\tlineCount = len(lines)\n",
    "\t\tif lineCount > 0:\n",
    "\t\t\tline = lines[0]  # take first only\n",
    "\t\t\tavg = avgFrom(line)\n",
    "\t\t\tprint(avg)\n",
    "\t\t\treturn True\n",
    "\t\t\t\n",
    "\t\treturn False\n",
    "\t\t\n",
    "\n",
    "patientCode = input('Enter patient''s code: ')\n",
    "fileName = buildName(patientCode)\n",
    "result = process(fileName)\n",
    "if process(fileName) == False:\n",
    "\tprint(False)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Ex 4.2\n",
    "\n",
    "In Your folder there are many text files containing data of surgeries. Those files are added and deleted when needed.\n",
    "Files are named with bed number ranging from 01 to 30.\n",
    "Write a program that reads existing files, prints file names and contents.\n",
    "\n",
    "(Program must be written using subprograms extensively.)\n",
    "\n",
    "\n",
    "Every file has the following structure:\n",
    "  name: John\n",
    "  surname: Doe\n",
    "  surgery_date:2020-11-21\n",
    "\n",
    "<b>Note:</b>\n",
    "    in some files you can have:\n",
    "\n",
    "surname: Smith\n",
    "name: Mark\n",
    "surgery_date:2020-11-23\n",
    "\n",
    "but program must:\n",
    "<ul>\n",
    "<li>write fields in the correct order</li>\n",
    "<li>replace numeric month with its English name.</li>\n",
    "</ul>\n",
    "\n",
    "\n",
    "Expected output:\n",
    "\n",
    "file name:  01.txt\n",
    "\n",
    "Doe John 2020-July-28 \n",
    "\n",
    "file name:  05.txt\n",
    "\n",
    "Smith Mark 2020-November-24\n",
    "\n",
    "\n",
    "(run note: You wuold see data for bed 1 and bed 5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "file name:  bed_01.txt\n",
      "Doe John 2020-July-28 \n",
      "file name:  bed_05.txt\n",
      "Smith Mark 2020-November-24 \n"
     ]
    }
   ],
   "source": [
    "MAX_BEDS = 10\n",
    "MONTHS = [\"January\", \"February\", \"March\",\"April\",\n",
    "\t\t  \"May\", \"June\", \"July\", \"August\",\n",
    "\t\t  \"September\", \"October\", \"November\", \"December\"]\n",
    "MONTHS_NUM = 12\n",
    "\n",
    "FILE_PREFIX = \"bed_\"\n",
    "FILE_SUFFIX = \".txt\"\n",
    "\n",
    "def buildName(bedNumber):\n",
    "\tbedNumberStr = str(bedNumber)\n",
    "\tif len(bedNumberStr) < 2:\n",
    "\t\tbedNumberStr = FILE_PREFIX + '0' + bedNumberStr\n",
    "\tfname = bedNumberStr + FILE_SUFFIX\n",
    "\treturn fname\n",
    "\n",
    "\n",
    "#if not found, return \"\"\n",
    "def findDataFor(fieldName, inLines):\n",
    "\tresult = \"\"\n",
    "\tl = len(inLines)\n",
    "\ti = 0\n",
    "\tfound = False\n",
    "\n",
    "\twhile i < l and not found:\n",
    "\t\tinfo = inLines[i].split(':')\n",
    "\t\tif info[0] == fieldName:\n",
    "\t\t\tfound = True\n",
    "\t\telse:\n",
    "\t\t\ti += 1\n",
    "\tif found:\n",
    "\t\tresult = info[1].strip()\n",
    "\treturn  result\n",
    "\n",
    "\n",
    "def convertDate(s):\n",
    "\tresult = \"invalid month\"\n",
    "\t#get month as index:\n",
    "\tyearStr = s[0:4]\n",
    "\tmonthStr = s[5:7]\n",
    "\tdayStr = s[8:10]\n",
    "\tif monthStr.isdigit():\n",
    "\t\tindex = int(monthStr)  - 1\n",
    "\t\tif index < MONTHS_NUM:\n",
    "\t\t\tmonthName = MONTHS[index]\n",
    "\n",
    "\tresult = yearStr + '-' + monthName + '-' + dayStr\n",
    "\treturn result\n",
    "\n",
    "def printPatientSurgeryData(lines):\n",
    "\t#use  array of field names to parse rows.\n",
    "\tkeys = ['surname',  'name', 'surgery_date']\n",
    "\tpatientRecord = \"\"\n",
    "\tfor k in keys:\n",
    "\t\ts = findDataFor(k, lines)\n",
    "\t\tif len(s)>0: # test if we cannot find a valid field.\n",
    "\t\t\tif k != \"surgery_date\" :\n",
    "\t\t\t\tpatientRecord = patientRecord + s + \" \"\n",
    "\t\t\telse:\n",
    "\t\t\t\tpatientRecord = patientRecord + convertDate(s) + \" \"\n",
    "\n",
    "\tprint(patientRecord)\n",
    "\n",
    "def process(fileName):\n",
    "\t# print('processing ', fileName)\n",
    "\ttry:\n",
    "\t\tf = open(fileName)\t\t\n",
    "\texcept:\n",
    "\t\treturn False\n",
    "\telse:\n",
    "\t\tprint('file name: ', fileName)\n",
    "\t\tlines = f.read().splitlines()\n",
    "\t\tprintPatientSurgeryData(lines)\n",
    "\t\tf.close()\n",
    "\t\t\n",
    "for bedNumber in range(1,MAX_BEDS):\n",
    "\tfileName = buildName(bedNumber)\n",
    "\tresult = process(fileName)\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Ex 4.3 Pandas dataframes\n",
    "\n",
    "\n",
    "Download from: \n",
    "https://www.kaggle.com/ronitf/heart-disease-uci\n",
    "\n",
    " the file: heart.csv.\n",
    "\n",
    "It contains the following information:\n",
    "1.  age \n",
    "2.  sex 1 = M, 0 = F)\n",
    "3.  chest pain type (4 values) \n",
    "4.  resting blood pressure \n",
    "5.  serum cholestoral in mg/dl \n",
    "6.  fasting blood sugar > 120 mg/dl\n",
    "7.  resting electrocardiographic results (values 0,1,2)\n",
    "8.  maximum heart rate achieved \n",
    "9.  exercise induced angina \n",
    "10. oldpeak = ST depression induced by exercise relative to rest \n",
    "11. the slope of the peak exercise ST segment \n",
    "12. number of major vessels (0-3) colored by flourosopy \n",
    "13. thal: 3 = normal; 6 = fixed defect; 7 = reversable defect\n",
    "\n",
    "example:\n",
    "    \n",
    " age,sex,cp,trestbps,chol,fbs,restecg,thalach,exang,oldpeak,slope,ca,thal,target\n",
    "\n",
    " 63,1,3,145,233,1,0,150,0,2.3,0,0,1,1\n",
    "\n",
    "\n",
    "\n",
    "Write a python program that reads that files and outputs \n",
    "the average age of females with chest pain (cp) <2 with angina during exercises.\n",
    "\n",
    "Or print False if cannot read it.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "ename": "ModuleNotFoundError",
     "evalue": "No module named 'pandas'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mModuleNotFoundError\u001b[0m                       Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-8-1b9a2ec0b3e8>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0;32mimport\u001b[0m \u001b[0mpandas\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0mpd\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      2\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      3\u001b[0m \u001b[0mFILE_NAME\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m\"heart.csv\"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      4\u001b[0m \u001b[0mFEMALE\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;36m0\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mModuleNotFoundError\u001b[0m: No module named 'pandas'"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "FILE_NAME = \"heart.csv\"\n",
    "FEMALE = 0\n",
    "\n",
    "try:\n",
    "    dataFrame = pd.read_csv(FILE_NAME)\n",
    "\n",
    "except:\n",
    "    print(False)\n",
    "\n",
    "else:\n",
    "    filtered = dataFrame[\n",
    "        (dataFrame['cp']<2) &\n",
    "        (dataFrame['sex'] == FEMALE) &\n",
    "        (dataFrame['exang'] == 1)\n",
    "        ]\n",
    "    #print(filtered)\n",
    "    average = filtered['age'].mean()\n",
    "    print(average)\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Ex 4.4 GMQL data manipulation with Dataset\n",
    "\n",
    "GMQL project (http://www.bioinformatics.deib.polimi.it/genomic_computing/GMQL/) contains a lot of genomic recordsets.\n",
    "\n",
    "\n",
    "Suppose You team has already downloaded from:\n",
    "\n",
    "http://www.gmql.eu/gmql-rest/\n",
    "\n",
    "\n",
    "A file named: queryname_20201120_192414_RESULT_DS.zip \n",
    "\n",
    "\n",
    "When expanded, you got this layout:\n",
    "\n",
    "S_00000.gdm\n",
    "\n",
    "S_00000.gdm.meta\n",
    "\n",
    "S_00001.gdm\n",
    "\n",
    "S_00001.gdm.meta\n",
    "\n",
    "S_00002.gdm\n",
    "\n",
    "S_00002.gdm.meta\n",
    "\n",
    "S_00003.gdm\n",
    "\n",
    "S_00003.gdm.meta\n",
    "\n",
    "S_00004.gdm\n",
    "\n",
    "S_00004.gdm.meta\n",
    "\n",
    "S_00005.gdm\n",
    "\n",
    "S_00005.gdm.meta\n",
    "\n",
    "S_00006.gdm\n",
    "\n",
    "S_00006.gdm.meta\n",
    "\n",
    "info.txt\n",
    "\n",
    "query.txt\n",
    "\n",
    "schema.xml\n",
    "\n",
    "vocabulary.txt\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "Inside info.txt You can read:\n",
    "\n",
    "Number of samples\t7\n",
    "\n",
    "\n",
    "SO you can iterate and read all files inside \"files:\" folder.\n",
    "\n",
    "\n",
    "(Suppose your python program will run inside the main folder)\n",
    "\n",
    "Every file is a \"TAB delimited\" with this structure:\n",
    "\n",
    "\n",
    "chr1\t99\t150\t*\tGMQL\tRegion\t4\t.\t.\t10\t7.0638\t-1\t-1\n",
    "\n",
    "chr1\t119\t180\t*\tGMQL\tRegion\t3\t.\t.\t24\t19.7648\t-1\t-1\n",
    "\n",
    "chr1\t219\t240\t*\tGMQL\tRegion\t8\t.\t.\t13\t11.2001\t-1\t-1\n",
    "\n",
    "\n",
    "\n",
    "<b>Write a program that open all files and prints the width (end-start)\n",
    "Where pValue > 19 using Pandas dataset.\n",
    "</b>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "ename": "ModuleNotFoundError",
     "evalue": "No module named 'pandas'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mModuleNotFoundError\u001b[0m                       Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-12-2889e6f72ed3>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0;32mimport\u001b[0m \u001b[0mpandas\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0mpd\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      2\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      3\u001b[0m \u001b[0mINFO_FILE_NAME\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m\"info.txt\"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      4\u001b[0m \u001b[0mN_OF_SAMPLES_KEY\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m\"Number of samples\"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mModuleNotFoundError\u001b[0m: No module named 'pandas'"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "INFO_FILE_NAME = \"info.txt\"\n",
    "N_OF_SAMPLES_KEY = \"Number of samples\"\n",
    "\n",
    "FILE_PREFIX = \"S_\"\n",
    "FILE_SUFFIX = \".gdm\"\n",
    "\n",
    "THRESHOLD = 0.19\n",
    "\n",
    "def readInfoFile():\n",
    "\ttry:\n",
    "\t\tdataFrame = pd.read_csv(INFO_FILE_NAME, sep = '\\t', header = None)\n",
    "\t\tresult = int(dataFrame.loc[5,1])\n",
    "\texcept:\n",
    "\t\tresult = 0\n",
    "\treturn result\n",
    "\n",
    "def process(nofFiles):\n",
    "\tfor i in range(0,nOfFiles):\n",
    "\t\tfname = buildFileName(i)\n",
    "\t\tprocessFile(fname)\n",
    "\n",
    "def buildFileName(index):\n",
    "\ts = FILE_PREFIX + '{:0>5}'.format(index)+FILE_SUFFIX\n",
    "\treturn s\n",
    "\n",
    "\n",
    "def processFile(fname):\n",
    "\tprint(fname)\n",
    "\tdataFrame = pd.read_csv(fname, sep = '\\t',  header = None)\n",
    "\n",
    "\tdataFrame.columns = ['chr', 'start', 'end', 'strand', 'source', 'feature',\n",
    "\t\t\t\t\t\t 'score', 'frame', 'name', 'signal',\n",
    "\t\t\t\t\t\t'pvalue', 'qvalue', 'peak'\n",
    "\t\t\t\t\t\t ]\n",
    "\n",
    "\tdataFrame['pvalue'] = pd.to_numeric(dataFrame['pvalue'])\n",
    "\n",
    "\tfiltered = dataFrame[dataFrame['pvalue']/100 > THRESHOLD]\n",
    "\n",
    "\twidth_df = filtered['end'] -  filtered['start']\n",
    "\tfor v in width_df:\n",
    "\t\tprint(v)\n",
    "\n",
    "nOfFiles = readInfoFile()\n",
    "if nOfFiles>0:\n",
    "\tprint(nOfFiles ,\" to be processed\")\n",
    "\tprocess(nOfFiles)\n",
    "else:\n",
    "\tprint(INFO_FILE_NAME + \" is missing\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Ex4.4 GMQL data manipulation with files functions\n",
    "\n",
    "Rewrite the above program using standard file functions making extensive use of subprograms.\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "7  to be processed\n",
      "S_00000.gdm\n",
      "61\n",
      "S_00001.gdm\n",
      "101\n",
      "112\n",
      "S_00002.gdm\n",
      "51\n",
      "S_00003.gdm\n",
      "31\n",
      "S_00004.gdm\n",
      "41\n",
      "S_00005.gdm\n",
      "61\n",
      "11\n",
      "S_00006.gdm\n",
      "51\n",
      "101\n",
      "51\n"
     ]
    }
   ],
   "source": [
    "INFO_FILE_NAME = \"info.txt\"\n",
    "N_OF_SAMPLES_KEY = \"Number of samples\"\n",
    "\n",
    "FILE_PREFIX = \"S_\"\n",
    "FILE_SUFFIX = \".gdm\"\n",
    "THRESHOLD = 0.19\n",
    "\n",
    "def readFileLines(fileName):\n",
    "\ttry:\n",
    "\t\tf = open(fileName)\n",
    "\texcept:\n",
    "\t\treturn []\n",
    "\telse:\n",
    "\t\tlines = f.read().splitlines()\n",
    "\t\treturn lines\n",
    "\n",
    "\n",
    "def readInfoFile():\n",
    "\tresult = 0\n",
    "\tlines = readFileLines(INFO_FILE_NAME)\n",
    "\t#scan for N_OF_SAMPLES_KEY:\n",
    "\tfor l in lines:\n",
    "\t\t(key, value) = l.split('\\t')\n",
    "\t\tif key ==  N_OF_SAMPLES_KEY:\n",
    "\t\t\tresult = int(value)\n",
    "\t\t\tbreak\n",
    "\n",
    "\treturn result\n",
    "\n",
    "def buildFileName(index):\n",
    "\ts = FILE_PREFIX + '{:0>5}'.format(index)+FILE_SUFFIX\n",
    "\treturn s\n",
    "\n",
    "\n",
    "def processLine(l):\n",
    "\t#print(l)\n",
    "\tparts = l.split('\\t')\n",
    "\t#print(len(parts))\n",
    "\tend = int(parts[2])\n",
    "\tstart = int(parts[1])\n",
    "\tpValue = float(parts[10])\n",
    "\twidth = end - start\n",
    "\tif pValue /100 > THRESHOLD:\n",
    "\t\tprint(width)\n",
    "\n",
    "def processFile(fname):\n",
    "\tprint(fname)\n",
    "\tlines = readFileLines(fname)\n",
    "\tlineCount = len(lines)\n",
    "\tif lineCount > 0:\n",
    "\t\tfor l in lines:\n",
    "\t\t\tprocessLine(l)\n",
    "\telse:\n",
    "\t\tprint(\"cannot open \"  + fname)\n",
    "\n",
    "def process(nofFiles):\n",
    "\tfor i in range(0,nOfFiles):\n",
    "\t\tfname = buildFileName(i)\n",
    "\t\tprocessFile(fname)\n",
    "#main:\n",
    "nOfFiles = readInfoFile()\n",
    "if nOfFiles>0:\n",
    "\tprint(nOfFiles ,\" to be processed\")\n",
    "\tprocess(nOfFiles)\n",
    "else:\n",
    "\tprint(INFO_FILE_NAME + \" is missing\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
